### PR TITLE
Correctly resolve tags for function overloads

### LINKED ISSUE #30181: Quick info returns wrong jsdoc tag for function overload
**TypeScript Version:**  3.4.0-dev.20190301

**Search Terms:**
- quickInfo
- jsdoc
- tags

**Code**

```ts
declare function foo(): void;

/**
 * @deprecated
 */
declare function foo(x: number): void;

foo()
```

1. Hover on `foo` on the last line

**Expected behavior:**
No deprecation tag shown since we are using the non deprecated

**Actual behavior:**
A deprecation tag is returned in the quickinfo response:

```
[Trace  - 5:08:21 PM] Response received: quickinfo (311). Request took 2 ms. Success: true 
Result: {
    "kind": "function",
    "kindModifiers": "declare",
    "start": {
        "line": 8,
        "offset": 1
    },
    "end": {
        "line": 8,
        "offset": 4
    },
    "displayString": "function foo(): void (+1 overload)",
    "documentation": "",
    "tags": [
        {
            "name": "deprecated"
        }
    ]
}
```

**Playground Link:** 

**Related Issues:**

### PR DESCRIPTION
**Pull request by jeanp413**
Fixes #30181 #35933

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
There's a common thing people do where they only add JSDoc to the first overload. My feeling is that we should first try to get JSDoc for the applicable overload, and if we don't find anything, we should try to find JSDoc for any other overload. Thoughts @mjbvz @RyanCavanaugh?

**Comment by ajafff** [comment]
I typically don't add jsdoc comments, but when I do, I use it to deprecate a single overload.

**Comment by DanielRosenwasser** [comment]
@ajafff please post future comments in the form of a meme if possible.

image

**Comment by sandersn** [comment]
@DanielRosenwasser That's a good point. I'm going to look at Definitely Typed.

So far, it looks like the numbers are really low compared to [compared to the total number of overloads](https://twitter.com/sanders_n/status/1141821637219586048). Admittedly, I'm only looking at d.ts and DT is relatively well-documented.

Some super early numbers:
Two-overload sets: 50 of 1,000,000 only document the first overload.
Three-overload sets: 25 of 5000.

I'll run over all of DT (and its dependencies) but right now I think this is a good change.

File: src/services/symbolDisplay.ts [hunk 1]
```diff
@@ -501,7 +501,7 @@ namespace ts.SymbolDisplay {
 
         if (!documentation) {
             documentation = symbol.getDocumentationComment(typeChecker);
```
Comment (sandersn) [inline src/services/symbolDisplay.ts]: I think you should add the same code here.

File: tests/cases/fourslash/quickInfoJsDocTagsFunctionOverload.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
+/// <reference path='fourslash.ts'/>
+
+// @Filename: quickInfoJsDocTagsFunctionOverload.ts
+////declare function /*1*/foo(): void;
+////
+/////**
+//// * @tag Tag text
+//// */
+////declare function /*2*/foo(x: number): void
```
Comment (sandersn) [inline tests/cases/fourslash/quickInfoJsDocTagsFunctionOverload.ts]: this test is a good start.

1. can you add tests for documentation as well?
2. can you add a test with the sequences [doc/no-tag, doc/tag], [doc/no-tag, doc/no-tag], [no-doc/no-tag, doc/tag], [no-doc/no-tag, doc/no-tag], [no-doc/no-tag, no-doc/tag]? There are other combinations but those are the ones that come to mind.
3. I couldn't get this to work in my editor when I played with it locally, although the tests pass. Probably I just set it up wrong, but you might want to make sure it at leasts works in your editor.

**Comment by sandersn** [comment]
Actually, looks like I was asking the wrong question. Turns out the number of first-jsdoc-only overloads is quite high, _of the overloads with jsdoc_:
Selection_045

The yellow line is overloads with any jsdoc; the red line is overloads with jsdoc only on the first signature. Blue is all overloads. X=size of overload set; Y=number of each size of set.

Looks like @DanielRosenwasser is right; the jsdoc for the first signature of the overload should be applied to all others that do not themselves have jsdoc. The same is not true for non-first signatures.

**Comment by jeanp413** [comment]
@sandersn Pushed a better fix and added more unit tests

File: tests/cases/fourslash/commentsOverloads.ts [hunk 3]
```diff
@@ -363,7 +363,7 @@ verify.signatureHelp({ marker: "38", overloadsCount: 2, docComment: "this is sig
 verify.quickInfoAt("38q", "var i3_i: i3\n(a: number) => number (+1 overload)", "this is signature 1");
 
 verify.signatureHelp({ marker: "39", overloadsCount: 2 });
-verify.quickInfoAt("39q", "var i3_i: i3\n(b: string) => number (+1 overload)");
+verify.quickInfoAt("39q", "var i3_i: i3\n(b: string) => number (+1 overload)", "this is signature 1");
```
Comment (sandersn) [inline tests/cases/fourslash/commentsOverloads.ts]: same: this documentation is supposed to be on `(a: number)` not `(b: string)`.

File: tests/cases/fourslash/commentsOverloads.ts [hunk 4]
```diff
@@ -428,7 +428,7 @@ verify.signatureHelp({ marker: "57", overloadsCount: 2, docComment: "c2 1" });
 verify.quickInfoAt("57q", "constructor c2(a: number): c2 (+1 overload)", "c2 1");
 
 verify.signatureHelp({ marker: "58", overloadsCount: 2 });
-verify.quickInfoAt("58q", "constructor c2(b: string): c2 (+1 overload)");
+verify.quickInfoAt("58q", "constructor c2(b: string): c2 (+1 overload)", "c2 1");
```
Comment (sandersn) [inline tests/cases/fourslash/commentsOverloads.ts]: same: `(a: number)` is the one with jsdoc, not `(b: string)`

File: tests/cases/fourslash/commentsOverloads.ts [hunk 5]
```diff
@@ -277,15 +277,15 @@ verify.completions(
...
+            { name: "f2", text: "function f2(a: number): number (+1 overload)" },
             { name: "f3", text: "function f3(a: number): number (+1 overload)" },
             { name: "f4", text: "function f4(a: number): number (+1 overload)", documentation: "this is signature 4 - with number parameter" },
         ],
     },
     {
         marker: "18",
         includes: [
-            { name: "i1_i", text: "var i1_i: i1\nnew (b: number) => any (+1 overload)" },
+            { name: "i1_i", text: "var i1_i: i1\nnew (b: number) => any (+1 overload)", documentation: "new 1" },
```
[the remark this answers is not in the document]
Comment (jeanp413) [reply tests/cases/fourslash/commentsOverloads.ts]: I took a look at this and it's working as intended following:
> the jsdoc for the first signature of the overload should be applied to all others that do not themselves have jsdoc

Is this an exception to the above rule?
Should this exception only be applied when the `useConstructSignatures === true`?

https://github.com/microsoft/TypeScript/blob/376acadc2df270014b808d832b0eee5f9e2fde65/src/services/symbolDisplay.ts#L222-L226

[the remark this answers is not in the document]
Comment (sandersn) [reply tests/cases/fourslash/commentsOverloads.ts]: No, I just missed that rule.

**Pull request merged**